binlog_streamer.go

```diff
@@ -151,42 +160,60 @@ func (s *BinlogStreamer) Run() {
 			continue
 		}
 
+		evPosition := mysql.Position{
+			Name: currentFilename,
+			Pos:  ev.Header.LogPos,
+		}
+
+		s.logger.WithFields(logrus.Fields{
+			"position":                   evPosition.Pos,
+			"file":                       evPosition.Name,
+			"type":                       fmt.Sprintf("%T", ev.Event),
+			"lastStreamedBinlogPosition": s.lastStreamedBinlogPosition,
+		}).Debug("reached position")
+
+		isEventPositionResumable := false
+		isEventPositionValid := true
+
 		switch e := ev.Event.(type) {
 		case *replication.RotateEvent:
-			// This event is needed because we need to update the last successful
-			// binlog position.
-			s.lastResumableBinlogPosition.Pos = uint32(e.Position)
-			s.lastResumableBinlogPosition.Name = string(e.NextLogName)
+			// This event is used to keep the "current binlog filename" of the binlog streamer in sync.
+			nextFilename = string(e.NextLogName)
+
+			isFakeRotateEvent := ev.Header.LogPos == 0 && ev.Header.Timestamp == 0
+			if isFakeRotateEvent {
+				// Sometimes the RotateEvent is fake and not a real rotation. we want to ignore the log position in the header for those events
+				// https://github.com/percona/percona-server/blob/3ff016a46ce2cde58d8007ec9834f958da53cbea/sql/rpl_binlog_sender.cc#L278-L287
+				// https://github.com/percona/percona-server/blob/3ff016a46ce2cde58d8007ec9834f958da53cbea/sql/rpl_binlog_sender.cc#L904-L907
+
+				// However, we can always advance our lastStreamedBinlogPosition according to its data fields
```

Comment:
>Is it possible for the stopAtBinlogPosition to ever be the "fake" RotateEvent?

No, because AFAICT, similar to our replication, the core MySQL replication client also doesn't advance its "master position" (the "last streamed position") based on these fake events